pkg/cloudprovider/cloudapi/azure/azure_security.go

```diff
@@ -344,100 +336,172 @@ func (computeCfg *computeServiceConfig) buildEffectiveNSGSecurityRulesToApply(ap
-			// skip any rules created by current processing appliedToGroup (as we have new rules for this group).
-			ruleAddrGroupName := desc.AppliedToGroup
+			// check if the rule is created by current processing appliedToGroup.
 			_, _, isNepheControllerCreatedRule := securitygroup.IsNepheControllerCreatedSG(desc.AppliedToGroup)
-			if isNepheControllerCreatedRule && strings.Compare(ruleAddrGroupName, appliedToGroupNepheControllerName) == 0 {
-				continue
+			if isNepheControllerCreatedRule && strings.Compare(desc.AppliedToGroup, appliedToGroupNepheControllerName) == 0 {
+				// skip the rule if found in remove list.
+				removeRules := rmIngressSecurityRules
+				if *rule.Properties.Direction == armnetwork.SecurityRuleDirectionOutbound {
+					removeRules = rmEgressSecurityRules
+				}
+				idx, found := findSecurityRule(rule, removeRules)
+				if found {
+					// remove the rule from remove list to avoid redundant comparison.
+					removeRules[idx] = nil
+					continue
+				}
 			}
 		}
 
-		// preserve rules.
+		// preserve current rules that do not need to update.
 		if *rule.Properties.Direction == armnetwork.SecurityRuleDirectionInbound {
-			currentNsgIngressRules = append(currentNsgIngressRules, *rule)
+			currentNsgIngressRules = append(currentNsgIngressRules, rule)
 		} else {
-			currentNsgEgressRules = append(currentNsgEgressRules, *rule)
+			currentNsgEgressRules = append(currentNsgEgressRules, rule)
 		}
 	}
 
-	agAsgMapByNepheControllerName, _, err := getNepheControllerCreatedAsgByNameForResourceGroup(computeCfg.asgAPIClient, rgName)
+	allIngressRules := updateSecurityRuleNameAndPriority(currentNsgIngressRules, addIngressSecurityRules)
+	allEgressRules := updateSecurityRuleNameAndPriority(currentNsgEgressRules, addEgressSecurityRules)
+
+	return append(allIngressRules, allEgressRules...), nil
+}
+
+// buildEffectivePeerNSGSecurityRulesToApply prepares the update rule cloud api payload from internal rules that require peering.
+func (computeCfg *computeServiceConfig) buildEffectivePeerNSGSecurityRulesToApply(appliedToGroupID *securitygroup.CloudResourceID,
+	addRules []*securitygroup.CloudRule, rmRules []*securitygroup.CloudRule, perVnetAppliedToNsgName string,
+	rgName string, ruleIP *string) ([]*armnetwork.SecurityRule, error) {
+	// get current rules for applied to SG azure NSG
+	nsgObj, err := computeCfg.nsgAPIClient.get(context.Background(), rgName, perVnetAppliedToNsgName, "")
 	if err != nil {
 		return []*armnetwork.SecurityRule{}, err
 	}
+	if nsgObj.Properties == nil {
+		return nil, fmt.Errorf("properties field empty in nsg object %s", *nsgObj.ID)
+	}
 
-	newIngressSecurityRules, err := convertIngressToPeerNsgSecurityRules(appliedToGroupID, ingressRules,
+	addIRule := make([]*securitygroup.CloudRule, 0)
+	rmIRule := make([]*securitygroup.CloudRule, 0)
+	addERule := make([]*securitygroup.CloudRule, 0)
+	rmERule := make([]*securitygroup.CloudRule, 0)
+	for _, rule := range addRules {
```

Comment:
move to new function.. GetRules, pass ingress as boolean to figure out direction

Reply:
Done